rust/datafusion/src/dataframe.rs

```diff
@@ -174,4 +174,18 @@ pub trait DataFrame {
 
     /// Return the logical plan represented by this DataFrame.
     fn to_logical_plan(&self) -> LogicalPlan;
+
+    /// Return a DataFrame with the explanation of its plan so far.
+    ///
+    /// ```
+    /// # use datafusion::prelude::*;
+    /// # use datafusion::error::Result;
+    /// # fn main() -> Result<()> {
+    /// let mut ctx = ExecutionContext::new();
+    /// let df = ctx.read_csv("tests/example.csv", CsvReadOptions::new())?;
+    /// let batches = df.limit(100)?.explain(false)?.collect()?;
+    /// # Ok(())
+    /// # }
+    /// ```
+    fn explain(&self, verbose: bool) -> Result<Arc<dyn DataFrame>>;
```

Comment:
I agree with your concern @jorgecarleitao and would expect `explain()` to just print the plan rather than return a DataFrame. That would be easy to implement (just call pretty_print there) but hard to test.

Reply:
Or we could add something like  `fn explain(&self, verbose: bool) -> String` if you wanted to keep it like spark. It would seem that to get a String for the explain plan, we would need to run the various optimizers and physical planner, but that is probably ok. 

I am not as familiar with Spark's DataFrame API so I do not have a good feel for how people would want this API to behave.

Reply:
So Spark has a couple different things to help people debug. `explain` literally prints the query plan(s) to stdout. It is also possible to access the underlying logical and physical query plans using `df.queryExecution.plan` or `df.queryExecution.executedPlan` for example. We already have `to_logical_plan` and I do think that we eventually should expose the physical plan as well, but not until we have made a decision on how we're implementing the physical plan enum and optimizer (there are JIRAs open for these issues already).

Reply:
I find it poor design that `.explain` prints directly to the stdout in spark. IMO saving 1 extra line (print) of code is not a sufficiently good reason to outright spam stdout and limit so much what a user can do with `.explain`.

Some downstream consequences of this decision in spark:
* it makes it much more difficult to log it correctly
* the popular pyspark can't use it to convert it to a Python string and prettify it when it is being used in notebooks

I agree with `fn explain(&self, verbose: bool) -> String` (prob. `Result<String>`). For a user, the difference is

```
df.explain()
```

vs 

```
println("{}", df.explain()?)
```

I find the latter more expressive of the user's intention, and gives them the freedom to pipe the result to whatever stream they want.

Reply:
I agree completely. Spark's design is flawed in this case.